src/runtime.rs

```diff
@@ -0,0 +1,371 @@
+            let sender = inner.as_ref().unwrap().clone();
+            sender
+        }
+
+        pub async fn spawn<F, Fut>(&self, f: F)
+        where
+            F: FnOnce() -> Fut + Send + 'static,
+            Fut: Future + 'static,
+        {
+            let f = || f().map(|_| ()).boxed_local();
+            self.sender()
+                .into_send_async(Task(Box::new(f)))
+                .await
+                .unwrap();
+        }
+
+        pub fn spawn_sync<F, Fut>(&self, f: F)
+        where
+            F: FnOnce() -> Fut + Send + 'static,
+            Fut: Future + 'static,
+            Fut::Output: 'static,
+        {
+            let f = || f().map(|_| ()).boxed_local();
+            self.sender().send(Task(Box::new(f))).unwrap();
+        }
+
+        pub async fn run<F, Fut>(&self, f: F) -> Fut::Output
+        where
+            F: FnOnce() -> Fut + Send + 'static,
+            Fut: Future + 'static,
+            Fut::Output: Send + 'static,
+        {
+            let (tx, rx) = futures::channel::oneshot::channel();
+            let f = || {
+                f().map(|x| {
+                    tx.send(x).ok();
+                })
+                .boxed_local()
+            };
+            self.sender()
+                .into_send_async(Task(Box::new(f)))
+                .await
+                .unwrap();
+            rx.await.unwrap()
+        }
+
+        pub fn run_sync<F, Fut>(&self, f: F) -> Fut::Output
+        where
+            F: FnOnce() -> Fut + Send + 'static,
+            Fut: Future + 'static,
+            Fut::Output: Send + 'static,
+        {
+            let (tx, rx) = std::sync::mpsc::channel();
+            let f = move || {
+                f().map(move |x| {
+                    // on completion, send result back to caller
+                    tx.send(x).ok();
+                })
+                .boxed_local()
+            };
+            // wait for completion (blocking)
+            self.sender().send(Task(Box::new(f))).unwrap();
+            rx.recv().unwrap()
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use std::{
+            collections::BTreeSet,
+            sync::{Arc, Mutex},
+        };
+        use tokio::task::LocalSet;
+        use tracing_subscriber::{prelude::*, EnvFilter};
+
+        fn thread_name() -> String {
+            std::thread::current().name().unwrap().to_string()
+        }
+
+        #[test]
+        fn tpc_runtime_smoke() {
+            tracing_subscriber::registry()
+                .with(tracing_subscriber::fmt::layer().with_writer(std::io::stderr))
+                .with(EnvFilter::from_default_env())
+                .try_init()
+                .ok();
+            let rt = Runtime::new("test", num_cpus::get());
+            let handle = rt.handle().clone();
+            let values = Arc::new(Mutex::new(BTreeSet::new()));
+            let values2 = values.clone();
+            rt.handle().run_sync(|| async move {
```

Comment:
I believe this is called `_blocking` in other runtimes, so maybe lets use that language instead of `_sync`?